File: src/main/csharp/p00/Problem0007.cs
using System;

namespace ProjectEuler.main.csharp.p00
{
    public class Problem0007
    {
        public static int Solve()
        {
            var count = 2;
            int i;
            for (i = 5; count < 10001; i++)
            {
                if ((i & 1) == 0) continue;

                var found = true;
                var sqrt = Math.Sqrt(i);
                for (var j = 3; j <= sqrt; j += 2)
                    if (i % j == 0)
                    {
                        found = false;
                        break;
                    }

                if (found) count++;
            }

            return i - 1;
        }
    }
}

File: src/main/kotlin/p00/Problem0007.kt
package p00

class Problem0007 {

    fun solve(): Int {
        tailrec fun helper(primes: List<Int>, possiblePrimes: List<Int>): List<Int> =
                when {
                    possiblePrimes.isEmpty() -> primes
                    else -> {
                        val prime = possiblePrimes.first()
                        val sievedPossiblePrimes = possiblePrimes.filter { it % prime != 0 }
                        helper(primes + prime, sievedPossiblePrimes)
                    }
                }

        return helper(emptyList(), (2..150000).toList())[10000]
    }

}